File: blockchain-101-java/src/main/java/code4fun/SImplestBlockchain.java
package code4fun;

import org.apache.commons.codec.digest.DigestUtils;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

/**
 * Created by max on 18-12-14.
 */
public class SImplestBlockchain {
    static class Transaction {
        private String fromAddress;
        private String toAddress;
        private BigDecimal amount;

        public Transaction(String fromAddress, String toAddress, BigDecimal amount) {
            this.fromAddress = fromAddress;
            this.toAddress = toAddress;
            this.amount = amount;
        }

        public String getFromAddress() {
            return fromAddress;
        }

        public void setFromAddress(String fromAddress) {
            this.fromAddress = fromAddress;
        }

        public String getToAddress() {
            return toAddress;
        }

        public void setToAddress(String toAddress) {
            this.toAddress = toAddress;
        }

        public BigDecimal getAmount() {
            return amount;
        }

        public void setAmount(BigDecimal amount) {
            this.amount = amount;
        }

        @Override
        public String toString() {
            return "Transaction{" +
                    "fromAddress='" + fromAddress + '\'' +
                    ", toAddress='" + toAddress + '\'' +
                    ", amount=" + amount +
                    '}';
        }
    }

    static class Block {
        private LocalDateTime timestamp;
        private List<Transaction> transactions;
        private String previousHash;
        private String hash = "";
        private int nonce = 0;

        public Block(LocalDateTime timestamp, List<Transaction> transactions,
                     String previousHash) {
            this.timestamp = timestamp;
            this.transactions = transactions;
            this.previousHash = previousHash;
        }

        public LocalDateTime getTimestamp() {
            return timestamp;
        }

        public void setTimestamp(LocalDateTime timestamp) {
            this.timestamp = timestamp;
        }

        public String getPreviousHash() {
            return previousHash;
        }

        public void setPreviousHash(String previousHash) {
            this.previousHash = previousHash;
        }

        public String getHash() {
            return hash;
        }

        public void setHash(String hash) {
            this.hash = hash;
        }

        public String calculateHash() {
            return DigestUtils.sha256Hex(toString());
        }

        public void mineBlock(int difficulty) {
            String prefix = IntStream.rangeClosed(1, difficulty)
                    .mapToObj(i -> "0").collect(Collectors.joining());
            while(!hash.startsWith(prefix)) {
                nonce ++;
                hash = calculateHash();
            }

            System.out.println("Block mined: " + hash);
        }

        @Override
        public String toString() {
            return "Block{" +
                    "timestamp=" + timestamp +
                    ", transactions=[" + transactions.stream()
                                            .map(Transaction::toString)
                                            .collect(Collectors.joining(", ")) +
                    "], previousHash='" + previousHash + '\'' +
                    ", nonce=" + nonce +
                    '}';
        }
    }

    static class Blockchain {
        private List<Block> chain = new ArrayList<>();
        private int difficulty = 4;
        private List<Transaction> pendingTransactions = new ArrayList<>();
        private BigDecimal miningReward = BigDecimal.valueOf(100);

        public Blockchain() {
            chain.add(createGenesisBlock());
        }

        private Block createGenesisBlock() {
            Block block = new Block(LocalDateTime.now(), Collections.emptyList(), "");
            block.setHash(block.calculateHash());
            return block;
        }

        public Block getLatestBlock() {
            return chain.get(chain.size() - 1);
        }

        public void minePendingTransactions(String miningRewardAddress) {
            Block block = new Block(LocalDateTime.now(), pendingTransactions, getLatestBlock().getHash());
            block.mineBlock(this.difficulty);
            pendingTransactions = new ArrayList<>();
            pendingTransactions.add(new Transaction("", miningRewardAddress, miningReward));
            System.out.println("Block successfully mined");
            chain.add(block);
        }

        public void createTransaction(Transaction transaction) {
            pendingTransactions.add(transaction);
        }

        public BigDecimal getBalanceOfAddress(String address) {
            return chain.stream().flatMap(block -> block.transactions.stream())
                    .map(transaction -> {
                        if (transaction.getFromAddress().equals(address)) {
                            return transaction.getAmount().multiply(BigDecimal.valueOf(-1));
                        }
                        if (transaction.getToAddress().equals(address)) {
                            return transaction.getAmount();
                        }
                        return BigDecimal.valueOf(0);
                    }).reduce(BigDecimal.ZERO, BigDecimal::add);

        }

        public boolean isChainValid() {
            for (int i = 1; i < chain.size(); i ++) {
                Block currentBlock = chain.get(i);
                Block previousBlock = chain.get(i - 1);

                if (!currentBlock.getHash().equals(currentBlock.calculateHash())) {
                    return false;
                }

                if (!currentBlock.getPreviousHash().equals(previousBlock.getHash())) {
                    return false;
                }
            }

            return true;
        }
    }

    public static void main(String[] args) {
        Blockchain blockchain = new Blockchain();
        blockchain.createTransaction(new Transaction("address1", "address2", BigDecimal.valueOf(100)));
        blockchain.createTransaction(new Transaction("address2", "address1", BigDecimal.valueOf(50)));
        System.out.println("Starting the miner..");
        blockchain.minePendingTransactions("miner1");
        System.out.println(String.format("Balance of the miner: %s is %s", "miner1", blockchain.getBalanceOfAddress("miner1").toString()));
        System.out.println(String.format("Balance of the miner: %s is %s", "address1", blockchain.getBalanceOfAddress("address1").toString()));
        System.out.println(String.format("Balance of the miner: %s is %s", "address2", blockchain.getBalanceOfAddress("address2").toString()));
        System.out.println("Starting the miner..");
        blockchain.minePendingTransactions("miner1");
        System.out.println(String.format("Balance of the miner: %s is %s", "miner1", blockchain.getBalanceOfAddress("miner1").toString()));
        System.out.println("Is chain valid: " + blockchain.isChainValid());
    }
}

File: simplest-blockchain.py
from hashlib import sha256
from datetime import datetime

class Transaction:
    def __init__(self, fromAddress, toAddress, amount):
        self.fromAddress = fromAddress
        self.toAddress = toAddress
        self.amount = amount

    def __str__(self):
        return f"Transaction{{fromAddress='{self.fromAddress}', toAddress='{self.toAddress}', amount={self.amount}}}"


class Block:
    def __init__(self, timestamp=datetime.now(), transactions=[], previousHash=''):
        self.timestamp = timestamp
        self.transactions = transactions
        self.previousHash = previousHash
        self.hash = ''
        self.nonce = 0       

    def __str__(self):
        transactionStrs = [t.__str__() for t in self.transactions]
        transactionsStr = ', '.join(transactionStrs)
        return f"Block{{timestamp={self.timestamp}, transactions=[{transactionsStr}], previousHash='{self.previousHash}', nonce={self.nonce}}}"

    def mineBlock(self, difficulty):
        prefex = '0' * difficulty
        while not self.hash.startswith(prefex):
            self.nonce += 1
            self.hash = self.calculateHash()
        print(f'Block mined: {self.hash}')

    def calculateHash(self):
        return sha256(self.__str__().encode('utf-8')).hexdigest()


class Blockchain:
    def __init__(self):
        self.chain = [self.createGenesisBlock()]
        self.difficulty = 4
        self.pendingTransactions = []
        self.miningReward = 100

    def createGenesisBlock(self):
        block = Block()
        block.hash = block.calculateHash()
        return block

    def getLatestBlock(self):
        return self.chain[-1]

    def minePendingTransactions(self, miningRewardAddress):
        block = Block(datetime.now(), self.pendingTransactions, self.getLatestBlock().hash)
        block.mineBlock(self.difficulty)
        self.pendingTransactions = [ Transaction('', miningRewardAddress, self.miningReward) ]
        print('Block successfully mined')
        self.chain.append(block)

    def createTransaction(self, transaction):
        self.pendingTransactions.append(transaction)

    def getBalanceOfAddress(self, address):
        balance = 0
        for block in self.chain:
            for transaction in block.transactions:
                if transaction.fromAddress == address:
                    balance -= transaction.amount
                elif transaction.toAddress == address:
                    balance += transaction.amount
        return balance

    def isChainValid(self):
        for i in range(1, len(self.chain)):
            currentBlock = self.chain[i]
            previousBlock = self.chain[i-1]

            if currentBlock.hash != currentBlock.calculateHash():
                return False

            if currentBlock.previousHash != previousBlock.hash:
                return False
        
        return True


blockchain = Blockchain()
blockchain.createTransaction(Transaction('address1', 'address2', 100))
blockchain.createTransaction(Transaction('address2', 'address1', 50))
print('Starting miner...')
blockchain.minePendingTransactions('miner1')
print(f"Balance of the miner: miner1 is {blockchain.getBalanceOfAddress('miner1')}")
print('Starting miner...')
blockchain.minePendingTransactions('miner1')
print(f"Balance of the miner: miner1 is {blockchain.getBalanceOfAddress('miner1')}")
print(f"Is chain valid: {blockchain.isChainValid()}")

block = blockchain.chain[1]
block.transactions = [ Transaction('some-address', 'my-address', 100000) ]
block.hash = block.calculateHash()
print(f"Is chain valid: {blockchain.isChainValid()}")